### PR TITLE
Fix output validation for LibTorch backend

### PR DESCRIPTION
**Pull request by CoderHam**
- [x] Extract model outputs only for the **Requested outputs** (Subset of those in model config).
- [x] Check **Datatype** of outputs after execution with that from model config.
- [x] Check **Shape** of outputs after execution with that from model config.
- [x] Check **Index** of outputs is valid for the number of outputs after execution.

### CODE REVIEW

File: src/backends/pytorch/libtorch_backend.cc [hunk 1]
```diff
@@ -613,12 +669,36 @@ LibTorchBackend::Context::Run(
...
   for (const auto& output : base->Config().output()) {
     const std::string& name = output.name();
     std::string index_str = name.substr(name.find(deliminator) + 2);
+    int op_index = std::atoi(index_str.c_str());
+    int max_index = outputs_.size() - 1;
+    if ((op_index < 0) || (op_index > max_index)) {
+      return Status(
+          RequestStatusCode::INVALID_ARG,
+          "The output " + name +
+              " in the model config refers to an output index which doesn't "
```
Comment (deadeyegoodwin) [inline src/backends/pytorch/libtorch_backend.cc]: Haven't we already tested (in the provider_utils) that the request asks only for outputs that are in the model config. That means that here we just have to make sure that every output required by the request is actually provided by the result. Isn't it ok if the result provides extra outputs that the request doesn't need (and/or that are listed in the model config). Perhaps unexpected? Perhaps we can log an error? But the request can still complete.

File: src/backends/pytorch/libtorch_backend.cc [hunk 2]
```diff
@@ -627,10 +707,17 @@ LibTorchBackend::Context::Run(
     // Checked at initialization time to make sure that STRING is not
     // being used for an output, so can just assume fixed-sized here.
     const DataType dtype = output_config->data_type();
+
+    // If a reshape is provided for the input then use that when
```
Comment (deadeyegoodwin) [inline src/backends/pytorch/libtorch_backend.cc]: input -> output

File: src/backends/pytorch/libtorch_backend.cc [hunk 3]
```diff
@@ -627,10 +707,17 @@ LibTorchBackend::Context::Run(
     // Checked at initialization time to make sure that STRING is not
```
Comment (deadeyegoodwin) [inline src/backends/pytorch/libtorch_backend.cc]: This comment belongs above the call to ReadFixedSizedOutputTensor below, not up here...

File: src/backends/pytorch/libtorch_backend.cc [hunk 4]
```diff
@@ -464,6 +509,17 @@ LibTorchBackend::Context::GetOutputTensor(
...
   try {
     torch::Tensor output_flat = (*outputs_)[op_index].flatten();
+
+    // verify output datatype matches datatype from model config
+    DataType rec_dtype = ConvertTorchTypeToDataType(output_flat.scalar_type());
+    if (dtype != rec_dtype) {
+      return Status(
+          RequestStatusCode::INVALID_ARG,
+          "unexpected datatype " + DataType_Name(rec_dtype) +
+              " for inference output 'OUTPUT__" + std::to_string(op_index) +
```
[the remark this answers is not in the document]
Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.cc]: You are correct. Will fix this

File: src/backends/pytorch/libtorch_backend.cc (hunk 1 above)
Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.cc]: Yes we test that the outputs asked for are present in the model config. However we do not test that the outputs have a valid *__\<index\> i.e. an the model config can have *_1 but the model can have only one output i.e. *__0. 
However now I see that these checks should be present in the earlier validation section as they need only be performed once. I will move it there.

File: src/backends/pytorch/libtorch_backend.cc [hunk 5]
```diff
@@ -613,12 +669,36 @@ LibTorchBackend::Context::Run(
...
   RETURN_IF_ERROR(Execute(&inputs_, &outputs_));
 
-  // Make sure each output is of the expected size and copy it into
-  // the payload responses.
+  // TODO Validate Output shape after model execution
+
   for (const auto& output : base->Config().output()) {
     const std::string& name = output.name();
     std::string index_str = name.substr(name.find(deliminator) + 2);
+    int op_index = std::atoi(index_str.c_str());
```
[the remark this answers is not in the document]
Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.cc]: Done

File: src/backends/pytorch/libtorch_backend.cc [hunk 6]
```diff
@@ -244,13 +277,19 @@ Status
...
     const auto pr = ConvertDataTypeToTorchType(io.data_type());
     if (!pr.first) {
       return Status(
           RequestStatusCode::INTERNAL,
           "unsupported datatype " + DataType_Name(io.data_type()) +
               " for output '" + io.name() + "' for model '" + name_ + "'");
+    } else {
+      const std::string& name = io.name();
+      int op_index = std::atoi(name.substr(name.find(deliminator) + 2).c_str());
+      io_index_map_[name] = op_index;
```
Comment (deadeyegoodwin) [inline src/backends/pytorch/libtorch_backend.cc]: I don't think you can use the same map for both input and output. What if I have an input with the same name as an output?

File: src/backends/pytorch/libtorch_backend.cc [hunk 7]
```diff
@@ -394,21 +433,52 @@ Status
...
+      outputs_, op_index, name, dtype, &content, &byte_size, &content_shape));
+
+  // verify shape of output matches shape from model config
+  const int batch_offset = ((max_batch_size_ == NO_BATCHING) ? 0 : 1);
+
+  for (int i = 0; i < dims.size(); i++) {
+    if (dims[i] != -1) {
+      if (dims[i] != content_shape[i + batch_offset]) {
+
+        // create print friendly array of shapes
```
Comment (deadeyegoodwin) [inline src/backends/pytorch/libtorch_backend.cc]: See src/core/model_config.h DimsListToString... you don't need to write your own versions.

File: src/backends/pytorch/libtorch_backend.cc (hunk 6 above)
Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.cc]: Right. Will fix that

File: src/backends/pytorch/libtorch_backend.cc [hunk 8]
```diff
@@ -226,13 +253,19 @@ Status
...
   for (const auto& io : ios) {
     const auto pr = ConvertDataTypeToTorchType(io.data_type());
     if (!pr.first) {
       return Status(
           RequestStatusCode::INTERNAL,
           "unsupported datatype " + DataType_Name(io.data_type()) +
               " for input '" + io.name() + "' for model '" + name_ + "'");
+    } else {
+      const std::string& name = io.name();
+      int op_index = std::atoi(name.substr(name.find(deliminator) + 2).c_str());
```
[the remark this answers is not in the document]
Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.cc]: Nope. Will throw an exception and add a unit test for this

File: src/backends/pytorch/libtorch_backend.cc (hunk 7 above)
Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.cc]: Got it

**Pull request merged**